openshift_tools/monitoring/generic_metric_sender.py

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python
+# vim: expandtab:tabstop=4:shiftwidth=4
+"""
+Base class for sending metrics
+"""
+
+import os
+import yaml
+
+
+class GenericMetricSenderException(Exception):
+    '''
+        GenericMetricSenderException
+        Exists to propagate errors up from the api
+    '''
+    pass
+
+
+class GenericMetricSender(object):
+    """
+    Abstract class that encapsulates all metric sending operations.
+    """
+
+    def __init__(self):
+        self.config = None
+        self.config_file = None
+
+    def parse_config(self):
+        """ parse default config file """
+        if not self.config:
+            if not os.path.exists(self.config_file):
+                raise GenericMetricSenderException(self.config_file + " does not exist.")
+            self.config = yaml.load(file(self.config_file))
+
+    def add_dynamic_metric(self, discovery_key, macro_string, macro_array, host=None, synthetic=False):
+        pass
```

Comment:
a. in hawkular all metrics are dynamic, you can invent new ids/items as you go.
`send_metrics(key='some.new.key.i.just.think.about', value=42, tags={...})` is ok.
b. the new keys can have tags:
b.1. from the `send_metrics` command it'self.
b.2. or from some `rule` in the config file, for example:
```
some.new.key.*:
   tags:
      type: 'node'
```
 
-------------------------

I think a nice algorithm will be:

a. use [optional] tags given by user when sending metrics:
    for example: `send_metric(key = 'cpu.usage', value = 2, tags = {type:'node', hostname: ... })`
b. use separate file to fill in missing tags if not given by the `send_metric` command.
    for example, if the config file has the lines:
```
cpu.usage:
  tags:
     units: 'cpus'
```
   the `send_metric` command will _also_ update the `units` tag